python/pyarrow/_substrait.pyx

```diff
@@ -15,35 +15,183 @@
 # specific language governing permissions and limitations
 # under the License.
 
+import base64
+import cloudpickle
+import inspect
+
 # cython: language_level = 3
-from cython.operator cimport dereference as deref
+from cython.operator cimport dereference as deref, preincrement as inc
 
+from pyarrow import compute as pc
 from pyarrow import Buffer
+from pyarrow.lib import frombytes, tobytes
 from pyarrow.lib cimport *
 from pyarrow.includes.libarrow cimport *
 from pyarrow.includes.libarrow_substrait cimport *
+from pyarrow._compute cimport FunctionRegistry
+
+
+from pyarrow._exec_plan cimport is_supported_execplan_output_type, execplan
+from pyarrow._compute import make_function_registry
+
+
+def make_extension_id_registry():
+    cdef:
+        shared_ptr[CExtensionIdRegistry] c_extid_registry
+        ExtensionIdRegistry registry
+
+    with nogil:
+        c_extid_registry = MakeExtensionIdRegistry()
+
+    return pyarrow_wrap_extension_id_registry(c_extid_registry)
+
+
+def _get_udf_code(func):
+    return frombytes(base64.b64encode(cloudpickle.dumps(func)))
+
+
+def get_udf_declarations(plan, extid_registry):
+    cdef:
+        shared_ptr[CBuffer] c_buf_plan
+        shared_ptr[CExtensionIdRegistry] c_extid_registry
+        vector[CUdfDeclaration] c_decls
+        vector[CUdfDeclaration].iterator c_decls_iter
+        vector[pair[shared_ptr[CDataType], c_bool]].iterator c_in_types_iter
+
+    c_buf_plan = pyarrow_unwrap_buffer(plan)
+    c_extid_registry = pyarrow_unwrap_extension_id_registry(extid_registry)
+    with nogil:
+        c_res_decls = DeserializePlanUdfs(
+            deref(c_buf_plan), c_extid_registry.get())
+    c_decls = GetResultValue(c_res_decls)
+
+    decls = []
+    c_decls_iter = c_decls.begin()
+    while c_decls_iter != c_decls.end():
```

Comment:
Hmm, normally, Cython allows you to iterate a `std::vector` the Python way:
```suggestion
    for c_decl in c_decls:
```
(you might have to `cdef c_decl` above)